generator/osm2type.cpp

```diff
@@ -319,15 +319,34 @@ class CachedTypes
 // e.g. highway-primary-bridge is left while highway-primary is removed.
 void LeaveLongestTypes(vector<generator::TypeStrings> & matchedTypes)
 {
-  auto const less = [](auto const & lhs, auto const & rhs) { return lhs > rhs; };
+  auto const equalPrefix = [](auto const & lhs, auto const & rhs)
+  {
+    size_t const prefixSz = std::min(lhs.size(), rhs.size());
+    bool const res = equal(lhs.begin(), lhs.begin() + std::min(size_t(2), prefixSz), rhs.begin());
+    if (res && prefixSz > 2)
+    {
+      // Define the set of exceptions like this:
+      if (lhs[1] == "charging_station")
+        return false;
+    }
+    return res;
+  };
 
-  auto const equals = [](auto const & lhs, auto const & rhs) {
-    if (rhs.size() > lhs.size())
-      return equal(lhs.begin(), lhs.end(), rhs.begin());
-    return equal(rhs.begin(), rhs.end(), lhs.begin());
+  auto const greaterBySize = [&equalPrefix](auto const & lhs, auto const & rhs)
+  {
+    if (equalPrefix(lhs, rhs))
+    {
+      if (lhs.size() != rhs.size())
+        return lhs.size() > rhs.size();
+
+      // - place-city-2 is better than place-city-4
```

Comment:
`place-city-10` will be above `place-city-2` or below?

Reply:
There will be only one and it will be place-city-2